KAFKA-9451: Update MockConsumer to support ConsumerGroupMetadata

mjsax opened the pull request:
Part of KIP-447

mjsax committed:
KAFKA-9441: Refactor TaskManager

mjsax commented:
Call for review @guozhangwang @abbccdda

mjsax commented on clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java:
```diff
@@ -563,7 +563,7 @@ private Long getEndOffset(List<Long> offsets) {
     @Override
     public ConsumerGroupMetadata groupMetadata() {
-        return null;
+        return new ConsumerGroupMetadata("dummy.group.id", 1, "1", Optional.empty());
```
We need this later (not related to this PR, but to follow up PR for KIP-447)

mjsax commented on clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java:
```diff
@@ -461,7 +461,7 @@ public synchronized void close() {
     }
 
-    @SuppressWarnings("deprecation")
+    @Deprecated
```
Side cleanup

mjsax commented on clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java:
```diff
@@ -308,7 +308,7 @@ private void verifyTransactionInFlight() {
         long offset = nextOffset(topicPartition);
         Completion completion = new Completion(offset, new RecordMetadata(topicPartition, 0, offset,
-                RecordBatch.NO_TIMESTAMP, Long.valueOf(0L), 0, 0), result, callback);
+                RecordBatch.NO_TIMESTAMP, 0L, 0, 0), result, callback);
```
Side cleanup

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -101,6 +101,7 @@
     private boolean commitNeeded = false;
     private boolean commitRequested = false;
 
+    @SuppressWarnings("rawtypes")
```
Side cleanup (multiple in the file)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -247,235 +241,6 @@ int getAssignmentErrorCode() {
         return assignmentErrorCode.get();
     }
 
-    static abstract class AbstractTaskCreator<T extends Task> {
```
We move this into `TaskManager`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -542,43 +305,21 @@ public static StreamThread create(final InternalTopologyBuilder builder,
         final Map<TaskId, Producer<byte[], byte[]>> taskProducers = new HashMap<>();
 
-        // TODO: refactor `TaskCreator` into `TaskManager`;
-        //  this will allow to reduce the surface area of `taskProducers` that is passed to many classes atm
```
This PR addresses this TODO

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -602,8 +343,6 @@ public static StreamThread create(final InternalTopologyBuilder builder,
             time,
             config,
-            activeTaskCreator.threadProducer,
-            taskProducers,
```
We don't need to pass the producers into `StreamThread` any longer now

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -686,14 +421,6 @@ private InternalConsumerConfig(final Map<String, Object> props) {
         }
     }
 
-    private static String getTaskProducerClientId(final String threadClientId, final TaskId taskId) {
```
moved to `TaskManager` (alse method below)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -1123,13 +850,11 @@ public final ThreadMetadata threadMetadata() {
+            state().name(),
+            getConsumerClientId(getName()),
+            getRestoreConsumerClientId(getName()),
+            taskManager.producerClientIds(getName()),
```
`producerClientIds()` contains the logic from above now

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -1139,11 +864,9 @@ StreamThread updateThreadMetadata(final String adminClientId) {
 
     private void updateThreadMetadata(final Map<TaskId, Task> activeTasks,
                                       final Map<TaskId, Task> standbyTasks) {
-        final Set<String> producerClientIds = new HashSet<>();
```
Getting `producerClientIds` is not moved into the TaskManager

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -1198,22 +919,7 @@ public String toString(final String indent) {
-            }
-        }
-        return result;
+        return taskManager.producerMetrics();
```
Logic move into `TaskManager`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -77,27 +316,127 @@
 
     private boolean rebalanceInProgress = false;  // if we are in the middle of a rebalance, it is not safe to commit
 
+    // for testing only
```
I added this constructor to allow mocking both task creators -- otherwise `TaskManagerTest` would require complex mocking and major rewrite

guozhangwang replied:
Not sure I follow here: can we just use easy-mock for the creators here, what are the actual issues that would cause major re-write in the test class?

mjsax replied:
If we cannot mock both `TaskCreators`, we need to mock the `InternalTopologyBuilder` what is much more complicated -- I tried this originally, but it was quite a mess. Also note, that `TaskManagerTest` makes a lot of assertions what method the `TaskManager` calls on the created tasks -- however, we would not be able to easily check this if we don't pass in mocked tasks, and we cannot pass in mock tasks without mocking the `TaskCreator`.

If you think this is too much of a hack, I would recommend to get rid of this constructor in a follow up PR and rewrite the `TaskManagerTest` their (to keep this PR focused and scoped).

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -77,27 +316,127 @@
+        log = logContext.logger(getClass());
+    }
+
+    TaskManager(final ChangelogReader changelogReader,
```
This is the actual constructor now

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -155,6 +155,7 @@ public void punctuate(final long timestamp) {
         }
     };
 
+    @SuppressWarnings("rawtypes")
```
side cleanup

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -1827,82 +1811,6 @@ private void verifyLogMessagesForSkippedRecordsForInvalidTimestamps(final LogCap
     }
 
-    @Test
-    public void shouldConstructProducerMetricsWithoutEOS() {
```
This and the next test are moved to `TaskManagerTest`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -49,10 +60,237 @@
+            this.cache = cache;
+            this.clientSupplier = clientSupplier;
+
+            this.createTaskSensor = ThreadMetrics.createTaskSensor(Thread.currentThread().getName(), streamsMetrics);
```
Minor change: We don't pass in `threadId` any longer but just call `Thread.currentThread.getName()` instead -- let me know what you think about it -- seems simpler to me to avoid passing around unnecessary parameters

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -49,10 +60,237 @@
+                storeChangelogReader,
+                time,
+                log);
+            createTaskSensor = ThreadMetrics.createTaskSensor(Thread.currentThread().getName(), streamsMetrics);
```
Similar to above: don't pass `threadId` any longer

mjsax commented:
Java 11:
```
org.apache.kafka.connect.mirror.MirrorConnectorsIntegrationTest.testReplication
kafka.admin.ReassignPartitionsClusterTest.shouldMoveSinglePartitionWithinBroker
```

Java 8: `kafka.api.SaslPlainSslEndToEndAuthorizationTest.testNoConsumeWithoutDescribeAclViaSubscribe`

No need to retest atm. Waiting for review.

mjsax commented on clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerGroupMetadata.java:
```diff
@@ -72,4 +72,21 @@ public String toString() {
             groupInstanceId.orElse(""));
     }
+
+    @Override
```
Need this to make `MockConsumerTest` pass.

guozhangwang reviewed (commented):
Thanks for the PR @mjsax, unfortunately its scope is overlapped with https://github.com/apache/kafka/pull/8213. The good news is that I think both of these two PRs are doing similar things here so with either one merged in, the other's scope can be largely reduced. Since @vvcephei 's PR is closer to be merged after the added unit tests, I'm thinking maybe we can merge that one first and rebase this one? Hopefully by rebasing many of the changes here can be saved as the creators are extracted out into new classes.

guozhangwang commented:
@mjsax could you rebase this PR so that we can review and merge soon?

mjsax commented:
@guozhangwang @abbccdda Rebased this PR. Call for review.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -304,37 +303,20 @@ public static StreamThread create(final InternalTopologyBuilder builder,
 
-        final ThreadCache cache = new ThreadCache(logContext, cacheSizeBytes, streamsMetrics);
-
-        final ActiveTaskCreator activeTaskCreator = new ActiveTaskCreator(
```
I think it's actually a better idea to keep extracting the active / standby task creators out of the constructor of task-manager, since by doing that in the tests we can have mocks for those two, and also we can avoid the extra constructors you added below. Thoughts?

Also cc @abbccdda @vvcephei who reviewed / worked on the class cleanups recently.

mjsax replied:
I guess @vvcephei refactoring did already the required work to get rid of passing around `threadProducer` and `taskProducers` references -- while I see the argument about testing and the second constructor, it still seems weird if the `StreamThread` that has nothing to do with tasks needs to create both task-creator objects -- it should be encapsulated in `TaskManager` -- maybe we can have some setters on TaskManager for testing instead of the second constructor (the corresponding members would not be `final` any longer for this case)

guozhangwang replied:
Logically it is more natural to have the task creation logic inside task-manager, yes; but the tricky part is that task creation involves creating a lot of its modules like state-manager and record-collector, so if we can pass them as parameters it is easier to mock them without creating those modules; so practically speaking I think it's better to always pass in those creators as parameters to task manager so that it's easier to mock, since there's really no real disadvantage of doing so right?

abbccdda replied:
I also don't feel strong about this refactoring. In the effort of moving task creator towards `TaskManager` we are also passing in more parameters than what `TaskManager` needs as well. I would recommend we reduce the PR to address https://issues.apache.org/jira/browse/KAFKA-9676 only, since we do miss unit test coverage for task creators.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreatorTest.java:
```diff
@@ -0,0 +1,155 @@
+    private ActiveTaskCreator activeTaskCreator;
+
+    @Test
+    public void shouldConstructProducerMetricsWithoutEOS() {
```
Thanks for the added test coverage!!

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -105,6 +105,8 @@
     private final TopicPartition t1p2 = new TopicPartition(topic1, 2);
     private final Set<TopicPartition> taskId02Partitions = mkSet(t1p2);
 
+    private final MockClientSupplier mockClientSupplier = new MockClientSupplier();
```
This seems not used.

mjsax replied:
Good catch -- missed to remove it after rebasing and moving the tests to `ActiveTaskCreatorTest`

guozhangwang reviewed (commented):
Made another pass, left some more comments.

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreatorTest.java:
```diff
@@ -0,0 +1,155 @@
+import static org.junit.Assert.assertEquals;
+
+@RunWith(EasyMockRunner.class)
+public class ActiveTaskCreatorTest {
```
Since we did this, could we also add test for `StandbyTaskCreator` to resolve https://issues.apache.org/jira/browse/KAFKA-9676?

abbccdda replied:
Could we add a test for standby?

mjsax commented:
Updated this PR.

This PR has nothing to do with KAFKA-9441 any longer. It does a small preparation for KAFKA-9451, hence, I think we should commit if with this ticket number?

I don't intend to address KAFKA-9679, even if I added some tests in this PR.

Thoughts?

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreatorTest.java:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    @Test
+    public void shouldConstructProducerMetricsWithEOS() {
```
Could we use `verify` on mocks for both test cases?

mjsax replied:
We don't really care what methods get called, do we? We are only interesting in the result of the method call -- if the internal implementation on how we build the metric change, the test should still pass.

abbccdda commented:
Overall the changes LGTM, just have 2 minor comments

abbccdda commented:
> I don't intend to address KAFKA-9679, even if I added some tests in this PR.

What do you mean? I thought this combines with TaskManagerTest shall be sufficient to address 9679? Otherwise I don't think this PR is very much useful to merge alone.

mjsax commented:
The original intent of this PR was to refactor -- however, the refactoring overlapped with John's PR (I was not aware that we has working on refactoring, too). Hence, I agree that the main purpose of the PR is void, as we also don't move the task creators into `TaskManager`. However, instead of just closing the PR, we can merge what we have to partly address 9679 -- but I really want to focus on kip-477 and not spent more time writing unit test that we missed to add when the new classed where added.

mjsax commented:
Java 8: `org.apache.kafka.connect.util.KafkaBasedLogTest.testSendAndReadToEnd`
Java 11: `kafka.admin.ReassignPartitionsClusterTest.shouldTriggerReassignmentWithZnodePrecedenceOnControllerStartup`

Retest this please

guozhangwang commented:
> The original intent of this PR was to refactor -- however, the refactoring overlapped with John's PR (I was not aware that we has working on refactoring, too). Hence, I agree that the main purpose of the PR is void, as we also don't move the task creators into `TaskManager`. However, instead of just closing the PR, we can merge what we have to partly address 9679 -- but I really want to focus on kip-477 and not spent more time writing unit test that we missed to add when the new classed where added.

Hey guys, I think you are talking about KAFKA-9676 right? :) Anyways, I think that's fair. I made another pass over it and it LGTM so @mjsax please feel free to merge as is after green. We just need to update the description / title of the commit message -- we can use KAFKA-9676 as the title and still leave the ticket open since as discussed above it is not intended to complete 9676.

abbccdda commented:
@mjsax @guozhangwang could we at least address KAFKA-9679 with just one more test case for `StandbyTaskCreator`, since we are getting so close here? If we don't even have time to complete that, I doubt why we would still prioritize merging this PR first as this is currently not related to KIP-447.

guozhangwang commented:
> @mjsax @guozhangwang could we at least address KAFKA-9679 with just one more test case for `StandbyTaskCreator`, since we are getting so close here? If we don't even have time to complete that, I doubt why we would still prioritize merging this PR first as this is currently not related to KIP-447.

I feel neutral about that: on the one hand, I agree the scope of this PR is not very relevant to 447, but on the other hand, completing KIP-447's development should still be top priority at the moment and this PR contains some cleanups for it so I think we do not necessarily need to hold this until 447 is done.

guozhangwang commented:
The failed test seems consistent in other PRs as well, I suspect the current trunk is also broken by this test.

mjsax commented:
Java8: `org.apache.kafka.connect.util.KafkaBasedLogTest.testSendAndReadToEnd`
Java11: `org.apache.kafka.connect.util.KafkaBasedLogTest.testSendAndReadToEnd`

Same test as above, even if it should be unrelated.

Retest this please.

abbccdda commented:
@guozhangwang Cool, I have done my part of the soft rejection :)

mjsax commented:
Rebased to resolve merge conflict. Jenkins should now pass as the failing test was fixed.

mjsax commented:
Rebased to resolve merge conflicts.

mjsax commented:
Java 8 passed.
Java 11: `kafka.api.ConsumerBounceTest.testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup`

Retest this please.

guozhangwang reviewed (approved):
Please feel free to merge afterwards.

mjsax commented:
Java 8 passed.
Java 11: `kafka.api.ConsumerBounceTest.testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup`

That is a different test. Merging this.
